### PR TITLE
[web] Expose browser_detection through ui_web.

### PR DESCRIPTION
**Pull request by ditman**
This PR moves the core of `browser_detection.dart` to a location in `dart:ui_web` so it can be used by apps and plugins.

In order for the code to be a little bit tidier in ui_web, it's encapsulated in a singleton instance that can be accessed through `BrowserDetection.instance` or a top level global `browser` in `dart:ui_web`.

## Issues

* Needed to fix: https://github.com/flutter/flutter/issues/128943
* Needed to land: https://github.com/flutter/flutter/pull/147346

## Tests

Updated affected tests. Mostly the update was to call the methods from `web_ui.browser.methodName` rather than a global scope. Also split the tests for this module in two files:

* `engine_browser_detect_test.dart` - with the tests specific to the engine (capability detection, etc...)
* `browser_detect_test.dart` - only the tests pertaining to the "core" of the library.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/lib/src/engine/browser_detection.dart [hunk 1]
```diff
@@ -3,220 +3,51 @@
...
 ///
-/// Taken from the navigator platform.
-/// <https://developer.mozilla.org/en-US/docs/Web/API/NavigatorID/platform>
-enum OperatingSystem {
-  /// iOS: <http://www.apple.com/ios/>
-  iOs,
-
-  /// Android: <https://www.android.com/>
-  android,
+/// Flutter web considers "mobile" everything that not [isDesktop].
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/browser_detection.dart:16]:
```suggestion
/// Flutter web considers "mobile" everything that's not [isDesktop].
```

File: lib/web_ui/lib/src/engine/browser_detection.dart [hunk 2]
```diff
@@ -3,220 +3,51 @@
...
-    return BrowserEngine.firefox;
-  }
-
-  // Assume Blink otherwise, but issue a warning.
-  print(
-      'WARNING: failed to detect current browser engine. Assuming this is a Chromium-compatible browser.');
-  return BrowserEngine.blink;
-}
+/// A flag to check if the current [operatingSystem] is a laptop/desktop
+/// operating system.
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/browser_detection.dart:11]: Might wanna keep it consistent with `isMobile`?
```suggestion
/// A flag to check if the current browser is running on a laptop/desktop device.
```

File: lib/web_ui/lib/src/engine/browser_detection.dart [hunk 3]
```diff
@@ -3,220 +3,51 @@
...
-  iOs,
-
-  /// Android: <https://www.android.com/>
-  android,
+/// Flutter web considers "mobile" everything that not [isDesktop].
+bool get isMobile => ui_web.browser.isMobile;
 
-  /// Linux: <https://www.linux.org/>
-  linux,
+/// Whether the current [browserEngine] is [BrowserEngine.blink] (Chrom(e|ium)).
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/browser_detection.dart:19]: You may have to use `[ui_web.browserEngine]` so dartdocs can link to it correctly. Same for `[ui_web.BrowserEngine.blink]`.

File: lib/web_ui/lib/src/engine/browser_detection.dart [hunk 4]
```diff
@@ -246,46 +77,38 @@ bool get isChrome110OrOlder {
 // since we check this on every frame.
 bool? _cachedIsChrome110OrOlder;
 
-/// If set to true pretends that the current browser is iOS Safari.
-///
-/// Useful for tests. Do not use in production code.
-@visibleForTesting
-bool debugEmulateIosSafari = false;
+/// Use in tests to simulated the detection of Chrome 110 or older on Windows.
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/browser_detection.dart:80]:
```suggestion
/// Used in tests to simulate the detection of Chrome 110 or older on Windows.
```

**Review by mdebbar — APPROVED** [review]

File: lib/web_ui/lib/src/engine/browser_detection.dart (hunk 3 above)
Comment (ditman) [reply lib/web_ui/lib/src/engine/browser_detection.dart]: Used "browser" instead of a bracket reference. Updated the value of the enum, though.

**Pull request merged**